### ISSUE TITLE
Training Code - New Features

### ISSUE THREAD

**Issue by melgor** [cmusatyalab/openface#106]
Adding new Features for Training Code, which will clean the code, reduce memory consumption during training and enable multiple-GPU.
Based on talk: https://groups.google.com/forum/?hl=en#!topic/cmu-openface/9Be_VMC659c
- [x] adding [cudnn.convert](https://github.com/soumith/cudnn.torch/blob/99883f3633ce8fc0567913a506203630da7ed95f/convert.lua#L32)
- [x] use [shareGradient](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Ffacebook%2Ffb.resnet.torch%2Fblob%2Fmaster%2Fmodels%2Finit.lua%23L42&sa=D&sntz=1&usg=AFQjCNGXZmtMQ-U6rrRIFpo4ZUKQg2vHog) idea from fb.resnet. Replaced by https://github.com/fmassa/optimize-net
- [x] support [multi-GPU](https://github.com/facebook/fb.resnet.torch/blob/master/models/init.lua#L89)
- [x] adding in-place activation to all nets (optnet take care about this)
- [x] adding [clearState](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Ftorch%2Fnn%2Fpull%2F526&sa=D&sntz=1&usg=AFQjCNF_lWecZ0I9AdRb6H9nR8dy9Aed0A)
- [ ] add [Transform class](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Ffacebook%2Ffb.resnet.torch%2Fblob%2Fmaster%2Fdatasets%2Ftransforms.lua&sa=D&sntz=1&usg=AFQjCNG5LcDlFxk7-EvK91DvRV2IcSbZAw) for Data Augumentation
- [x] speedup forward pass in [Tripelt-Criterion](https://github.com/Atcold/torch-TripletEmbedding/pull/3)
- [ ] speedup choosing triplets by calculating distance matrix
- [ ] speedup updating gradient by using single state per network (currently every layer have own state)

**Comment by melgor** [comment]
I tried to add "cudnn.convert". Unfortunately, there is a problem with GPU memory consumption. I was using it in my other projects and it works fine. 
I tested it and it is connected with 
`model:float()`
Without it it works fine...

**Comment by fmassa** [comment]
@melgor About the `shareGradient` part, I've started working on a generic code that allows to reuse the intermediary buffers of the network without having to define the sharing by hand https://github.com/fmassa/optimize-net
There are unit tests for common architectures, and I'll soon finish a version which works for training as well (there is an open PR for that already which works, just need to update the README and add some more tests).
I'll also soon add a more robust option for clearing the `gradWeights` when in inference mode (current simple approach loses the sharings if there are some), but that's not difficult to add.

**Comment by melgor** [comment]
I was trying to add muli-GPU stuff to the code but I failed. I have similar problem like in adding cuDNN. This is exacly about converting model to float before saving. 
https://github.com/cmusatyalab/openface/blob/master/training/train.lua#L91
Do we really need it? For sure everybody will use CUDA for training. We it would be better to provide the script which will convert it to float. 
But this will add other problem: we would not able to use any "clearState" solution, because it will break memory optimization made by 'optnet'

To sum up:
1. It would be easier to save CUDa model. 
2. Saved model will be bigger than now (as we save gradient matrices etc.)
2. It will need additional script for "clearState" and "float".

What to you think about it, @bamos ?

**Comment by bamos** [comment]
Hi @melgor - it seems reasonable to save the CUDA model to disk and then use a script that can convert it to float and clearing the state if it makes adding these features easier.

I have some issues running out of disk space so it's not ideal to use a larger model here. I wonder how much disk space compressing the model with something like https://groups.google.com/forum/#!topic/torch7/7hCpmhv20KY will save.

-Brandon.
